pkg/autoscaling/calculation.go

```diff
@@ -52,89 +49,399 @@ var (
+	if err != nil {
+		return "", err
+	}
+	if len(resp.Kvs) == 0 {
+		return "", errors.New(fmt.Sprintf("length of the response values of the key %s is 0", prometheusAddressKey))
 	}
 
-	if len(instances) == 0 {
-		return nil
+	address := &Address{}
+	err = json.Unmarshal(resp.Kvs[0].Value, address)
+	if err != nil {
+		return "", err
+	}
+
+	return address.String(), nil
+}
+
+func getTiDBPlans(querier Querier, instances []instance, strategy *Strategy, resourceMap map[string]uint64) ([]*Plan, error) {
+	return getCPUPlans(querier, instances, strategy, resourceMap, TiDB)
+}
+
+func getCPUPlans(querier Querier, instances []instance, strategy *Strategy, resourceMap map[string]uint64, component ComponentType) ([]*Plan, error) {
+	if strategy.Rules[0].CPURule == nil {
+		return nil, nil
 	}
 
 	now := time.Now()
-	totalCPUUseTime, err := getTotalCPUUseTime(querier, component, instances, now, MetricsTimeDuration)
+	// get cpu used times
+	cpuUsedTimes, err := querier.Query(NewQueryOptions(component, CPUUsage, now, MetricsTimeDuration))
 	if err != nil {
-		log.Error("cannot get total CPU used time", errs.ZapError(err))
-		return nil
+		return nil, err
 	}
+	// get cpu quotas
+	cpuQuotas, err := querier.Query(NewQueryOptions(component, CPUQuota, now, MetricsTimeDuration))
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug("get cpu usage information completed",
+		zap.String("component", component.String()),
+		zap.Any("cpuUsedTimes", cpuUsedTimes),
+		zap.Any("cpuQuotas", cpuQuotas),
+	)
+
+	var (
+		totalCPUUsedTime float64
+		totalCPUQuota    float64
+		cpuUsageLowNum   uint64
+	)
+
+	// get cpu threshold
+	cpuMaxThreshold, cpuMinThreshold := getCPUThresholdByComponent(strategy, component)
+	cpuUsageHighMap := make(map[float64]float64)
+	for instanceName, cpuUsedTime := range cpuUsedTimes {
+		cpuQuota, ok := cpuQuotas[instanceName]
+		if !ok {
+			// if corresponding cpu quota does not exist, consider this instance is in the low usage status,
+			// this could be useful to avoid scaling out incorrectly after scaling in heterogeneous groups
+			cpuUsageLowNum++
+			continue
+		}
+		cpuUsedTime /= MetricsTimeDuration.Seconds()
+		totalCPUUsedTime += cpuUsedTime
+		totalCPUQuota += cpuQuota
+		cpuUsage := cpuUsedTime / cpuQuota
 
-	currentQuota, err := getTotalCPUQuota(querier, component, instances, now)
-	if err != nil || currentQuota == 0 {
-		log.Error("cannot get total CPU quota", errs.ZapError(err))
-		return nil
+		if cpuUsage > cpuMaxThreshold {
+			cpuUsageHighMap[cpuUsage] = cpuQuota
+			continue
+		}
+		if cpuUsage < cpuMinThreshold {
+			cpuUsageLowNum++
+		}
 	}
 
-	totalCPUTime := float64(currentQuota) / milliCores * MetricsTimeDuration.Seconds()
-	usage := totalCPUUseTime / totalCPUTime
-	maxThreshold, minThreshold := getCPUThresholdByComponent(strategy, component)
+	totalInstanceCount := uint64(len(instances))
+	totalCPUUsage := totalCPUUsedTime / totalCPUQuota
+	cpuUsageTarget := (cpuMaxThreshold + cpuMinThreshold) / 2
```

Comment:
Is it reasonable?

Reply:
which one? sorry, I don't follow

Reply:
`cpuUsageTarget` is calculated through `cpuMaxThreshold` and `cpuMinThreshold`

Reply:
it's reansonable.
cpuUsageTarget is the target cpu usage percents that we want tikv/tidb instances stays,
`(cpuMaxThreshold + cpuMinThreshold) / 2`, it will return the middle value between `cpuMaxThreshold` and `cpuMinThreshold`, this value is neither too big nor too small.